### DISCUSSION TITLE
[Deployment] Deploying UUPS Upgradeable contracts and proxies via factories (Permissionless II Hackathon)

### DISCUSSION THREAD

**Discussion by scotthconner** [zkSync-Community-Hub/zksync-developers#129] [Deployments] [answered]
### Environment

Testnet

### zkSolc Version

v1.3.14

### zksync-web3 Version

^0.14.3

### Hardhat.config.ts

https://github.com/LocksmithWallet/smartrust-zksync/blob/main/hardhat.config.js

### Deployment Script (WITHOUT PRIVATE KEY)

(Not super relevant)

https://github.com/LocksmithWallet/smartrust-zksync/blob/main/tasks/genie.js

### Contract Code

https://github.com/LocksmithWallet/smartrust-zksync/blob/main/contracts/agents/KeyAddressFactory.sol

### Does this work on other EVMs? (If yes, please list at least 1 of them)

Ethereum, OPStack, Scroll, Polygon

### Description of What Your Contract Does

I have an Account Factory that deploys a UUPS Proxy account and its proxy in a single transaction as part of a smart contract. The create code provided in the documentation (using yul) does not produce an address that has a code length after return.

It seems as if the lifecycle of the bytecode is different for zkSync, as deploying the ERC1967 proxy fails to deploy due to the given address not being a contract (code.length == 0).

### Repo Link (Optional)

https://github.com/LocksmithWallet/smartrust-zksync

### Additional Details

The code length at the address after the create yul is empty.

**Comment by dutterbutter** [comment]
Hey hey @scotthconner 👋 

I can't recall where we left off in Austin re this issue but if you are still facing the issue please refer to the `@hardhat-zksync-upgradable` documentation that discusses UUPS upgradability [here](https://era.zksync.io/docs/tools/hardhat/hardhat-zksync-upgradable.html#uups-proxies).

**Reply by scotthconner** [reply]
I successfully use that just fine. The issue is deploying an implementation contract from within a factory results in a zero code length for the resulting address.

The behavior is divergent from other EVMs. This is happening within the VM, and not during deployment from a client using the utility.

Is there a solidity example of deploying UUPS proxies?

**Reply by dutterbutter** [reply]
There is no working example currently but we will create one and share with the community. That is a great suggestion.

**Reply by AlbionaHoti** [reply, marked as answer]
Hi @scotthconner, [we have an example](https://github.com/matter-labs/era-contracts/blob/main/zksync/contracts/bridge/L2ERC20Bridge.sol#L143-L161) that deploys a Beacon proxy on our code. Let me know if this helps your case.

**Comment by jakeaukuso** [comment]
It seems like there might be an issue with the bytecode lifecycle or deployment process specific to zkSync, causing the ERC1967 proxy deployment to fail due to the generated address not having any code (code.length == 0). This could be related to how zkSync handles bytecode or contract deployment. Further investigation and potential adjustments to the deployment process may be necessary to resolve this issue.

**Comment by kakarukjoseph** [comment]
It appears that the bytecode lifecycle in zkSync may differ from other EVMs, causing the deployment of the ERC1967 proxy to fail due to an empty code length at the given address. This discrepancy might require adjustments in your deployment process or bytecode generation specifically tailored for zkSync. Consider consulting zkSync's documentation or community for insights on bytecode handling and deployment in zkSync.

**Comment by tumb2ibonsa** [comment]
The provided smart contract aims to deploy a UUPS Proxy account and its proxy in a single transaction. However, when using zkSync, the bytecode generated by the create operation in Yul does not result in a contract address with non-zero code length, causing deployment failure. Further investigation is needed to understand the bytecode lifecycle on zkSync and address this issue.

**Comment by dyksterhouset** [comment]
Environment: Testnet
zkSolc Version: v1.3.14
zksync-web3 Version: ^0.14.3
Description: The bytecode lifecycle differs in zkSync, causing deployment failure of ERC1967 proxy due to empty code length at the address.

**Comment by anggabarber** [comment]
It seems like there is an issue with deploying an ERC1967 proxy contract on zkSync testnet using the provided create code in the documentation. The address returned by the create code does not have a code length, which causes the deployment to fail. It appears that the lifecycle of the bytecode may be different on zkSync compared to other EVMs. Without further information or access to the code repository, it is difficult to diagnose the exact issue.
